### DISCUSSION TITLE
Show timecode in timeline

### DISCUSSION THREAD

**Discussion by axiomgraph** [friction2d/friction#58] [Ideas]
In Enve, there was an option for showing seconds in the timeline, which did not work. It would be great if the timeline could show seconds and minutes.

**Comment by rodlie** [comment]
The feature was never implemented in enve (only the menu entry, which I removed).

If it's not too much work this is something I could add, personally I prefer to work with frames.

**Comment by AlexKiryanov** [comment]
Hi all! It would be great!

**Comment by axiomgraph** [comment]
For new users, it might be helpful.

**Comment by rodlie** [comment]
So...

![Screenshot from 2023-08-19 13-50-41](https://github.com/friction2d/friction/assets/34516798/7316ed9f-1cd6-4c3f-aded-e059d3972c55)

There are of course some issues, the timecode takes up a lot of space, the frames at the far left/right will always be cut of.

And:

* Now we also need to flip the bottom scrollbar to timecode and keep them in sync
* Now users expect the first/current/last frame widgets to also adapt to timecode, and a lot of converting to and from timecode will happen

I don't know ...

**Comment by AlexKiryanov** [comment]
Well, if I understood everything correctly, then probably for seconds it is necessary that the signatures change relative to the scale of the timeline.
When the scale is small, then sign only seconds, 1, 2, 3, etc. without 00:00:00:00:00
But I suspect it's too much to redo?
![image](https://github.com/friction2d/friction/assets/65244629/bd340cdc-4ee2-43d7-b5e3-75f21a56d7b3)

**Comment by AlexKiryanov** [comment]
![image](https://github.com/friction2d/friction/assets/65244629/fb3cf5c7-80af-4ffb-b07a-62161cf30527)
I am adding.
The scale adapts the values depending on the scale.
Zooming out changes the divisions.

**Comment by rodlie** [comment]
I don't think there is a need to modify the timecodes based on scale. It only a matter of aligning them (at least for the bottom scrollbar).

Current implementation:

## Zoomed in

![Screenshot from 2023-08-19 15-24-05](https://github.com/friction2d/friction/assets/34516798/5b881959-4403-439f-8965-c3bc90aca482)

Looks ok to me.

## Zoomed out
![Screenshot from 2023-08-19 15-24-15](https://github.com/friction2d/friction/assets/34516798/2d75b71c-7632-4dcf-8e11-43eaa2c153f3)

I don't have a solution for the top scrollbar yet (first/last is cut of), does this matter at all?

**Comment by AlexKiryanov** [comment]
"I don't have a solution for the top scrollbar yet (first/last is cut of), does this matter at all?"
Well, I don't think it's scary at all. Let it be cut off.

**Comment by rodlie** [comment]
I added the initial support for timecodes in `main`.

**Comment by axiomgraph** [comment]
one doubt ? do we need this area(filled with red)

![261800018-2d75b71c-7632-4dcf-8e11-43eaa2c153f3](https://github.com/friction2d/friction/assets/68183694/efbf674d-399c-47eb-8d3b-94171c959183)

**Reply by rodlie** [reply]
Yeah, it's mostly just for information and some padding for the small scroll bar.

**Comment by rodlie** [comment]
Something like this:

https://github.com/friction2d/friction/assets/34516798/e28bff01-034e-4e70-8ce4-ace450a65905

**Reply by axiomgraph** [reply]
this looks good. waiting for the release!

**Comment by rodlie** [comment]
Added some extra usability...

https://github.com/friction2d/friction/assets/34516798/d9afccc8-7ce7-4bd1-a5e3-ffdbeff6c03e

Scrolling on the timeline now supports modifiers `CTRL` and `SHIFT`.

* `CTRL` + wheel will now scale in/out the frame range as if you used the scrollbar
* `SHIFT` + wheel will now pan the frame range as if you used the scrollbar
* Without a modifier it works as before, scroll up/down

**Reply by axiomgraph** [reply]
That looks good too. 

Here are other things you should consider in the future roadmap:
Selective playback instead of playing everything from start to finish. It will improve workflow. faster loading.
![261800018-2d75b71c-7632-4dcf-8e11-43eaa2c153f3](https://github.com/friction2d/friction/assets/68183694/3a080722-57a1-4e92-bc03-45c490654023)
Marker + comment, improved motion blur The original Enve motion blur does not look as it should.
audio wave preview

**Reply by rodlie** [reply]
All valid points.

* In/Out points (based on selection and/or widget) would be the easiest to implement, maybe something for the next release (0.9.5)
* Motion blur: Someone need to make a new or fix the existing shader, contributions welcome!
* Markers will require a lot of work, don't expect this anytime soon (after v1)
* Audio wave preview would probably require the most work, but a valid feature for a future release (after v1)

**Comment by eFe-Munoz** [comment]
I also find it important to be able to work in seconds. It could be a solution to use the Cinema 4D system? it is still showing the frames, but you see in a box the corresponding second. Allows you to know what second you are in and to move to an exact second by typing it in. I share image:
![c4d](https://github.com/friction2d/friction/assets/92168582/f67d32e2-fd17-4c23-9b2c-8de463ca2d8d)

**Comment by rodlie** [comment]
https://github.com/friction2d/friction/assets/34516798/b19e469d-4555-49cd-8243-bbeb9090bb75


https://github.com/friction2d/friction/assets/34516798/e0f8a2f8-cade-403a-849d-6eee4252e996

**Comment by rodlie** [comment]
and very basic markers

https://github.com/friction2d/friction/assets/34516798/5585904c-1079-4c61-960f-0fec12e5f49b

**Comment by AlexKiryanov** [comment]
Hi all! Looks good!

**Comment by rodlie** [comment]
work-in-progress, but getting there :) Will do a v1 alpha-ish build in a week or so, just to get some feedback on the changes etc.

*It also a bit refreshing to do just ui stuff, doing fixes etc gets boring quick.*

**Comment by rodlie** [comment]
I will close this thread as we now have a dedicated issue https://github.com/friction2d/friction/issues/194

**Comment by eFe-Munoz** [comment]
Great work!!!
